src/froidure-pin-base.cpp

```diff
@@ -0,0 +1,725 @@
+    },
+    py::arg("batch_size"),
+    R"pbdoc(
+    Set a new value for the batch size.
+
+    :param batch_size: the new value for the batch size.
+    :type batch_size: int
+    The *batch size* is the number of new elements to be found by any call to
+    :any:`run` . This is used by, for example, :any:`FroidurePin::position` so
+    that it is possible to find the position of an element after only partially
+    enumerating the semigroup.The default value of the batch size is ``8192``.
+
+    :exceptions: This function is ``noexcept`` and is guaranteed never to
+    throw.
+
+    :complexity: Constant.
+
+    .. seealso::  :any:`batch_size`.
+
+
+    :returns: A reference to ``self``.
+
+    :rtype: FroidurePinBase
+    )pbdoc");*/
+
+    thing.def(
+        "current_normal_forms",
+        [](FroidurePinBase const& self) {
+          return py::make_iterator(self.cbegin_current_normal_forms(),
+                                   self.cend_current_normal_forms());
+        },
+        R"pbdoc(
+Returns an iterator containing the so-far enumerated normal forms (if any).
+This function returns an iterator pointing at the normal form of the first
+element of the semigroup represented by a :any:`FroidurePinBase` instance (if
+any). This function does not perform any enumeration of the :any:`FroidurePin`.
+If you want to obtain the complete set of rules, then use :any:`normal_forms`
+instead.
+
+:returns:
+  An iterator yielding a :any:`List[int]`.
+:rtype:
+  Iterator
+
+:complexity:
+  Constant.
+)pbdoc");
+
+    thing.def(
+        "current_rules",
+        [](FroidurePinBase const& self) {
+          return py::make_iterator(self.cbegin_current_rules(),
+                                   self.cend_current_rules());
+        },
+        R"pbdoc(
+Returns an iterator containing the so-far enumerated rules.
+Returns an iterator containing the rules in a confluent
+terminating rewriting system defining a semigroup isomorphic to the one
+defined by ``self``. This function does not perform any enumeration of the
+:any:`FroidurePin` object. If you want to obtain the complete set of rules,
+then use :any:`rules` instead.
+
+:returns:
+    An iterator yielding ``Tuple[List[int],List[int]]``.
+:rtype: Iterator
+
+:complexity: Constant
+
+.. TODO change the next line to doctest when FroidurePin available
+.. code-block:: c++
+
+    >>> S = FroidurePin(
+    ... BMat8([[1,  0,  0,  0],
+    ...        [1,  0,  0,  0],
+    ...        [1,  0,  0,  0],
+    ...        [1,  0,  0,  0]]),
+    ... BMat8([[0,  1,  0,  0],
+    ...        [0,  1,  0,  0],
+    ...        [0,  1,  0,  0],
+    ...        [0,  1,  0,  0]]),
+    ... BMat8([[0,  0,  1,  0],
+    ...        [0,  0,  1,  0],
+    ...        [0,  0,  1,  0],
+    ...        [0,  0,  1,  0]]),
+    ... BMat8([[0,  0,  0,  1],
+    ...        [0,  0,  0,  1],
+    ...        [0,  0,  0,  1],
+    ...        [0,  0,  0,  1]]))
+    >>> S.size()
+    4
+    >>> list(S.rules())
+)pbdoc");
```

Comment:
Should there be output here?

Reply:
`FroidurePinBase` is an abstract class, and I haven't yet completed `FroidurePin` so the output is pending :)